Split clonal report (#58)
immcantation/enchantr: inst/rstudio/templates/project/repertoire_analysis_project_files/index.Rmd

  skip_abundance_diversity <- FALSE
}

```

```{r}
# set empty default abundance
a <- new("AbundanceCurve")
```

```{r abundancebysample, results='asis', eval=isTRUE(any(isHeavyChain(db$locus))) && !isTRUE(skip_abundance_diversity)}

# calculate the rank-abundance curve
a <- estimateAbundance(
  db_keep_for_abundance %>%
    filter(isHeavyChain(locus)),
  group = "sample_id", min_n = params$min_n
)
if (nrow(a@abundance) == 0) {
  cat("\nAll groups failed to pass the threshold min_n=", params$min_n, ". Skipping clonal abundance report.\n\n")
}
```

```{r, child=c('abundance.Rmd'), eval=nrow(a@abundance)>0}
```

```{r}
# check whether to calculate clonal overlap
cloneby_sample_count <- db %>% group_by(.data[[params$cloneby]]) %>% summarize(n_samples=n_distinct(sample_id))

calculate_overlap <- any(cloneby_sample_count$n_samples>1)
max_n_samples <- max(cloneby_sample_count$n_samples)
fig_size <- min(8,max(4,0.7*max_n_samples))
text_size <- floor(min(3,20/max_n_samples)) # No text in heatmap cell if max_n_samples > 20
```

```{r clonal_overlap, eval=calculate_overlap, results='asis',fig.width=fig_size+1, fig.height=fig_size}
caption<- "Clonal Overlap Among Samples"

> cat("# Clonal Overlap Among Samples\n")

Reviewer: The procedure of bootstrapping and the reasoning behind it should be explained in the upper sections of the report, where we use it for the first time for clonal abundance.

if (text_size>0){
  cat("\n Heatmap of clonal overlap among samples. Numbers in each cell are the counts of clones shared by the x-axis sample and the y-axis sample. Percentages indicate the fraction of shared clones in the y-axis sample. Each cell is colored by percentage of clonal overlap. \n")
} else{

>     cat("\n Heatmap of clonal overlap among samples. Each cell is colored by percentage of clonal overlap, which is the fraction of shared clones in the y-axis sample. \n")

Reviewer: ```suggestion
    cat("\n Clonal overlap among samples. Each cell is colored by percentage of clonal overlap, which is the fraction of shared clones in the y-axis sample. \n")
```

  eeplot(
    clonal_overlap_p,
    outdir = params$outdir,
    file   = paste0("Clonal_Overlap_", params$cloneby, "_", item),
    caption = caption
  )
  
  htmltools::div(
    ggplotly(clonal_overlap_p, tooltip = c("sample_x", "sample_y", "overlap"))%>%
    layout(width = fig_size+1,height = fig_size),
    htmltools::h2("\n\n")
  )
})

# output all widgets + headings
htmltools::tagList(plot_list)

```

```{r}
# check whether to calculate bootstrapping clonal overlap
bootstrapping_cloneby_sample_count <- db %>% 
  filter(sample_id %in% unique(a@bootstrap$sample_id)) %>%       group_by(.data[[params$cloneby]]) %>%  
  summarize(n_samples=n_distinct(sample_id))

calculate_bootstrapping_overlap <- any(bootstrapping_cloneby_sample_count$n_samples>1)
max_n_samples <- max(bootstrapping_cloneby_sample_count$n_samples)
fig_size <- min(8,max(4,0.7*max_n_samples))
text_size <- floor(min(3,20/max_n_samples)) # No text in heatmap cell if max_n_samples > 20
```

```{r bootstrapping_clonal_overlap, eval=calculate_bootstrapping_overlap, results='asis',fig.width=fig_size+1, fig.height=fig_size}
cat("# Bootstrapping Clonal Overlap Among Samples\n")

> cat(paste0("\n Heatmap of average bootstrapping clonal overlap among samples. Each cell is colored by the percentage of average bootstrapping clonal overlap in the y-axis sample. Bootstrapping drew ",unique(a@n), " sequences from each sample each time for ", a@nboot, " bootstrapping repetitions. \n"))

Reviewer: ```suggestion
cat(paste0("\n Average clonal overlap among samples over all bootstrapping repetitions. Each cell is colored by the percentage of the average clonal overlap over the bootstrapping repetitions in the y-axis sample. Bootstrapping drew ",unique(a@n), " sequences from each sample each time for ", a@nboot, " bootstrapping repetitions. \n"))
```

  skip_abundance_diversity <- nrow(db_keep_for_abundance) == 0
  if(skip_abundance_diversity){
      cat("Samples have only one clone. Skip clone abundance and diversity. \n")
  }else{
      cat("Samples with only one clone are removed from clone abundance and diversity analysis. \n")
    }
} else {
  db_keep_for_abundance <- db
  skip_abundance_diversity <- FALSE
}

```

```{r}
# set empty default abundance
a <- new("AbundanceCurve")
```

```{r abundancebysample, results='asis', eval=isTRUE(any(isHeavyChain(db$locus))) && !isTRUE(skip_abundance_diversity)}

# calculate the rank-abundance curve
a <- estimateAbundance(
  db_keep_for_abundance %>%
    filter(isHeavyChain(locus)),
  group = "sample_id", min_n = params$min_n
)
if (nrow(a@abundance) == 0) {
  cat("\nAll groups failed to pass the threshold min_n=", params$min_n, ". Skipping clonal abundance report.\n\n")
}
```

```{r, child=c('abundance.Rmd'), eval=nrow(a@abundance)>0}
```

# Clonal overlap

<div class="introduction_paragraph">
Clonal overlap analysis computes the number of clones that are present across any pair of samples within a cloneby group. The cloneby group for this anlaysis was `r params$cloneby`. 

> In clonal overlap without bootstrapping, overlapping clones are counted using all sequences from each paired samples.

Reviewer: ```suggestion
We perform clonal overlap computations without and with bootstrapping. In clonal overlap without bootstrapping, overlapping clones are counted using all sequences from each paired samples, so it will show the actual number of clones that are present in both samples.
```

In clonal overlap with bootstrapping, the average number of overlapping clones is calculated across bootstrap resamples, which corrects for differences in sequencing depth among samples.
</div>

```{r}
# check whether to calculate clonal overlap
cloneby_sample_count <- db %>% group_by(.data[[params$cloneby]]) %>% summarize(n_samples=n_distinct(sample_id))

calculate_overlap <- any(cloneby_sample_count$n_samples>1)
max_n_samples <- max(cloneby_sample_count$n_samples)
fig_size <- min(8,max(4,0.7*max_n_samples))
text_size <- floor(min(3,20/max_n_samples)) # No text in heatmap cell if max_n_samples > 20

if (!calculate_overlap){
  cat("Clonal overlap analysis was not performed since there is only one sample per cloneby group")
}
```

```{r clonal_overlap, eval=calculate_overlap, results='asis',fig.width=fig_size+1, fig.height=fig_size}
caption<- "Clonal Overlap Among Samples"

cat("## Clonal overlap without bootstrapping\n")

if (text_size>0){

>   cat("\n Clonal overlap among samples. Numbers in each cell are the number of clones shared by the x-axis sample and the y-axis sample. Percentages indicate the fraction of shared clones from the total number of clones in the y-axis sample. Each cell is colored by percentage of clonal overlap. \n")

Reviewer: ```suggestion
  cat("\n Clonal overlap among samples considering all sequences in each sample.")
```

db_bootstrapping <- a@bootstrap

calculate_overlap_list <- bootstrapping_cloneby_sample_count %>% 
                             filter(n_samples>1) %>% 
                             pull(.data[[params$cloneby]])

# build a list of htmlwidgets
plot_list <- lapply(calculate_overlap_list, function(item) {

  samples <- clone_sizes %>% 
    filter(.data[[params$cloneby]]==item)  %>%
     pull(sample_id) %>%
     unique()
  
  bootstrapping_clonal_overlap_p <- Bootstrapping_ClonalOverlap_plot(
    db_bootstrapping %>% filter(sample_id %in% samples),
    geom_text_size = text_size
  )
  
  bootstrapping_clonal_overlap_p <- bootstrapping_clonal_overlap_p + ggtitle(paste0("Bootstrapping Clonal Overlap Heatmap(",params$cloneby, ": ", item,")"))

  eeplot(
    bootstrapping_clonal_overlap_p,
    outdir = params$outdir,
    file   = paste0("Bootstrapping_Clonal_Overlap_", params$cloneby, "_", item),
    caption = caption
  )
  
  htmltools::div(
    ggplotly(bootstrapping_clonal_overlap_p, tooltip = c("sample_x", "sample_y", "overlap"))%>%
    layout(width = fig_size+1,height = fig_size),
    htmltools::h2("\n\n")
  )
  }
)

# output all widgets + headings
htmltools::tagList(plot_list)

```

> ```{r eval=any(db$locus %in% c("IGH", "IGL", "IGK")), results='asis'}

Author: In addition, do you want V gene usage or V family usage or both in the report?
